File: libnd4j/include/ops/declarable/generic/sparse/csr_spmv_semiring.cpp
/* ******************************************************************************
 *
 *
 * This program and the accompanying materials are made available under the
 * terms of the Apache License, Version 2.0 which is available at
 * https://www.apache.org/licenses/LICENSE-2.0.
 *
 *  See the NOTICE file distributed with this work for additional
 *  information regarding copyright ownership.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations
 * under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 ******************************************************************************/

//
// CSR sparse matrix-vector product under a semiring: y = A ⊗ x
//
// Inputs:
//   [0] values  – 1D [nnz], float dtype   (non-zero values of A)
//   [1] colIdx  – 1D [nnz], INT32/INT64   (column indices)
//   [2] rowPtr  – 1D [rows+1], same INT   (row pointer array)
//   [3] x       – dense 1D vector, length = cols
// IArgs:
//   [0] rows
//   [1] cols
//   [2] semiring  (0=PLUS_TIMES, 1=MIN_PLUS, 2=MAX_PLUS, 3=OR_AND, 4=MIN_TIMES)
// Output:
//   [0] y dense 1D, length = rows
//

#include <system/op_boilerplate.h>
#if NOT_EXCLUDED(OP_csr_spmv_semiring)

#include <ops/declarable/headers/sparse_semiring.h>
#include <ops/declarable/helpers/sparse_blas_semiring.h>

namespace sd {
namespace ops {

CUSTOM_OP_IMPL(csr_spmv_semiring, 4, 1, false, 0, 3) {
  auto values = INPUT_VARIABLE(0);
  auto colIdx = INPUT_VARIABLE(1);
  auto rowPtr = INPUT_VARIABLE(2);
  auto x      = INPUT_VARIABLE(3);

  const LongType rows     = INT_ARG(0);
  const LongType cols     = INT_ARG(1);
  const int      semiring = static_cast<int>(INT_ARG(2));

  REQUIRE_TRUE(values->rankOf() == 1, 0,
               "csr_spmv_semiring: values must be 1D, got rank %d", values->rankOf());
  REQUIRE_TRUE(colIdx->rankOf() == 1, 0,
               "csr_spmv_semiring: colIdx must be 1D, got rank %d", colIdx->rankOf());
  REQUIRE_TRUE(rowPtr->rankOf() == 1, 0,
               "csr_spmv_semiring: rowPtr must be 1D, got rank %d", rowPtr->rankOf());
  REQUIRE_TRUE(x->rankOf() == 1, 0,
               "csr_spmv_semiring: x must be 1D, got rank %d", x->rankOf());
  REQUIRE_TRUE(values->lengthOf() == colIdx->lengthOf(), 0,
               "csr_spmv_semiring: values and colIdx must have the same length");
  REQUIRE_TRUE(rowPtr->lengthOf() == rows + 1, 0,
               "csr_spmv_semiring: rowPtr length must equal rows+1=%lld, got %lld",
               (long long)(rows + 1), (long long)rowPtr->lengthOf());
  REQUIRE_TRUE(colIdx->dataType() == rowPtr->dataType(), 0,
               "csr_spmv_semiring: colIdx and rowPtr must have the same integer dtype");
  REQUIRE_TRUE(semiring >= 0 && semiring <= 4, 0,
               "csr_spmv_semiring: semiring must be in [0,4], got %d", semiring);

  auto y = OUTPUT_NULLIFIED(0);

  sd::ops::helpers::csr_spmv_semiring(*values, *colIdx, *rowPtr, *x, *y, rows, cols, semiring);

  return sd::Status::OK;
}

DECLARE_SHAPE_FN(csr_spmv_semiring) {
  auto values = INPUT_VARIABLE(0);

  const LongType rows = INT_ARG(0);

  auto dtype    = values->dataType();
  auto outShape = ConstantShapeHelper::getInstance().createShapeInfo(dtype, 'c', {rows});
  return SHAPELIST(outShape);
}

DECLARE_TYPES(csr_spmv_semiring) {
  getOpDescriptor()
      ->setAllowedInputTypes(0, {ALL_FLOATS})  // values
      ->setAllowedInputTypes(1, {ALL_INTS})    // colIdx
      ->setAllowedInputTypes(2, {ALL_INTS})    // rowPtr
      ->setAllowedInputTypes(3, {ALL_FLOATS})  // x
      ->setAllowedOutputTypes({ALL_FLOATS});
}

}  // namespace ops
}  // namespace sd

#endif  // NOT_EXCLUDED(OP_csr_spmv_semiring)

File: nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/api/ops/impl/sparse/CsrSpmvSemiring.java
/*
 *  ******************************************************************************
 *  *
 *  *
 *  * This program and the accompanying materials are made available under the
 *  * terms of the Apache License, Version 2.0 which is available at
 *  * https://www.apache.org/licenses/LICENSE-2.0.
 *  *
 *  * See the NOTICE file distributed with this work for additional
 *  * information regarding copyright ownership.
 *  * Unless required by applicable law or agreed to in writing, software
 *  * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 *  * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 *  * License for the specific language governing permissions and limitations
 *  * under the License.
 *  *
 *  * SPDX-License-Identifier: Apache-2.0
 *  *****************************************************************************
 */

package org.nd4j.linalg.api.ops.impl.sparse;

import org.nd4j.autodiff.samediff.SDVariable;
import org.nd4j.autodiff.samediff.SameDiff;
import org.nd4j.linalg.api.buffer.DataType;
import org.nd4j.linalg.api.ndarray.INDArray;
import org.nd4j.linalg.api.ops.DynamicCustomOp;

import java.util.Collections;
import java.util.List;

/**
 * Forward-only CSR SpMV with a user-selected algebraic semiring.
 * Graph algorithms are not differentiable; {@link #doDiff} is intentionally absent.
 *
 * <p>C++ op name: {@code csr_spmv_semiring}
 * <ul>
 *   <li>Inputs:  values[nnz], colIdx[nnz INT32], rowPtr[rows+1 INT32], x[cols]</li>
 *   <li>IArgs:   rows, cols, semiring (integer code from {@link Semiring#code()})</li>
 *   <li>Output:  y[rows]</li>
 * </ul>
 *
 * <p>For each row i, the output is computed as:
 * <pre>
 *   y[i] = REDUCE_{k in row i}( add(acc, mul(values[k], x[colIdx[k]])) )
 * </pre>
 * where {@code mul} and {@code add} are determined by the chosen {@link Semiring}.
 *
 * @see Semiring
 * @see CsrSpmv
 */
public class CsrSpmvSemiring extends DynamicCustomOp {

    /** Number of rows in the sparse matrix. */
    private long rows;
    /** Number of columns in the sparse matrix. */
    private long cols;
    /** The algebraic semiring used for the multiply-reduce. */
    private Semiring semiring;

    /** No-arg constructor required for op-registry reflection. */
    public CsrSpmvSemiring() {}

    /**
     * Eager (INDArray) constructor.
     *
     * @param values   1D [nnz] non-zero values (floating dtype)
     * @param colIdx   1D [nnz] INT32 column indices
     * @param rowPtr   1D [rows+1] INT32 row pointers
     * @param x        1D dense vector of length {@code cols}
     * @param rows     number of rows in the logical dense matrix
     * @param cols     number of columns in the logical dense matrix
     * @param semiring the algebraic semiring to use for the multiply-reduce
     */
    public CsrSpmvSemiring(INDArray values, INDArray colIdx, INDArray rowPtr, INDArray x,
                            long rows, long cols, Semiring semiring) {
        super(new INDArray[]{values, colIdx, rowPtr, x}, null);
        this.rows = rows;
        this.cols = cols;
        this.semiring = semiring;
        addIArgument(rows, cols, semiring.code());
    }

    /**
     * SameDiff (symbolic) constructor.
     *
     * @param sd       the SameDiff graph
     * @param values   SDVariable [nnz] non-zero values
     * @param colIdx   SDVariable [nnz] INT32 column indices
     * @param rowPtr   SDVariable [rows+1] INT32 row pointers
     * @param x        SDVariable dense vector of length {@code cols}
     * @param rows     number of rows
     * @param cols     number of columns
     * @param semiring the algebraic semiring to use for the multiply-reduce
     */
    public CsrSpmvSemiring(SameDiff sd, SDVariable values, SDVariable colIdx, SDVariable rowPtr,
                            SDVariable x, long rows, long cols, Semiring semiring) {
        super(sd, new SDVariable[]{values, colIdx, rowPtr, x});
        this.rows = rows;
        this.cols = cols;
        this.semiring = semiring;
        addIArgument(rows, cols, semiring.code());
    }

    @Override
    public String opName() { return "csr_spmv_semiring"; }

    @Override
    public List<DataType> calculateOutputDataTypes(List<DataType> dataTypes) {
        // Output dtype matches values (input[0]) dtype
        return Collections.singletonList(dataTypes.get(0));
    }

    @Override
    public List<SDVariable> doDiff(List<SDVariable> grads) {
        throw new UnsupportedOperationException(
            "csr_spmv_semiring is forward-only: semiring ops (min-plus / max-plus / or-and) " +
            "are non-differentiable and cannot be used in automatic differentiation.");
    }
}
